File: ios/Classes/FlutterDevicePlugin.swift
import Flutter
import UIKit

public class FlutterDevicePlugin: NSObject, FlutterPlugin {
    let mead = Deeml()
    let cmpk = CMPK()
    let csele = CSele()
    let mels = MELS()
//    let photoUtil = PhotoUtil()
    let cal = CAll()
//    let calendarUtil = CalendarUtil()
    public static func register(with registrar: FlutterPluginRegistrar) {
        let cel = FlutterMethodChannel(name: "flutter_device", binaryMessenger: registrar.messenger())
        let deml = FlutterDevicePlugin()
        registrar.addMethodCallDelegate(deml, channel: cel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        format.dateFormat = "yyyy-MM-dd HH:mm:ss"
        
        switch call.method {
        case "device_id":
            let id = mead.id()
            result(id)
        case "install_referrer":
            result(mead.reffer())
        case "device_info":
            let data = mlea(from: mead.damldw())
            result(data)
        case "camera_picker":
            let option = call.arguments as? [String: Any?] ?? [:]
            cmpk.selecter(option: FJDAL(wamld: option["maxWidth"] as? Double, haldcl: option["maxHeight"] as? Double, qualmd: option["imageQuality"] as! CGFloat, fnoda: option["front"] as! Bool)) { res in
                if(res.code == ResultError.resultOK){
                    result(res.data)
                } else {
                    result(FlutterError(code: "\(res.code)", message: res.message, details: nil))
                }
            }
        case "contact_picker":
            csele.selecter{ res in
                if res.code == ResultError.resultOK {
                    result(res.data)
                } else {
                    result(FlutterError(code: "\(res.code)", message: res.message, details: nil))
                }
            }
            
        case "package_info":
            result(mead.myAp())
        case "position":
            mels.select{ res in
                if res.code == ResultError.resultOK {
                    result(res.data)
                } else {
                    result(FlutterError(code: "\(res.code)", message: res.message, details: nil))
                }
            }
//        case "app_list":
//            result("[]")
//        case "sms_list":
//            result("[]")
//        case "call_log_list":
//            result("[]")
//        case "photo_list":
//            photoUtil.getPhotos{ res in
//                if res.code == ResultError.resultOK {
//                    result(getJsonString(from: res.data))
//                }else {
//                    result(FlutterError(code: "\(res.code)", message: res.message, details: nil))
//                }
//            }
        case "contact_list":
            cal.findAll{ res in
                if res.code == ResultError.resultOK {
                    result(mlea(from: res.data))
                } else {
                    result(FlutterError(code: "\(res.code)", message: res.message, details: nil))
                }
            }
//        case "calendar_list":
//            calendarUtil.getContacts { res in
//                if res.code == ResultError.resultOK {
//                    result(getJsonString(from: res.data))
//                } else {
//                    result(FlutterError(code: "\(res.code)", message: res.message, details: nil))
//                }
//            }
        case "clean_preferences":
//            let defaults = UserDefaults.standard
//            defaults.removeObject(forKey: "app_timestamp")
            result(true)
        case "save_preferences":
            result(true)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/com/qc/flutter_device/FlutterDevicePlugin.kt
package com.qc.flutter_device

import androidx.activity.ComponentActivity
import androidx.lifecycle.lifecycleScope
import com.google.gson.Gson
import com.qc.device.DataCenter
import com.qc.device.model.ResultError
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.launch

/** FlutterDevicePlugin */
class FlutterDevicePlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var activity: ComponentActivity
    private lateinit var channel: MethodChannel
    private lateinit var contactPicker: ContactPicker
    private lateinit var cameraPicker: CameraPicker
    private lateinit var dataCenter: DataCenter

    private var GSON = Gson()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_device")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {

        when (call.method) {
            "device_id" -> {
                val id = dataCenter.getDeviceId()
                result.success(id)
            }

            "install_referrer" -> {
                dataCenter.getReferrer {
                    if (it.code == ResultError.RESULT_OK) {
                        result.success(GSON.toJson(it.data))
                    } else {
                        result.error(it.code.toString(), it.message, null)
                    }
                }
            }

            "contact_picker" -> {
                contactPicker.picker {
                    result.success(GSON.toJson(it.data))
                }
            }

            "camera_picker" -> {
                val arg = call.arguments as Map<*, *>
                val imageOption = ImageOption(
                    maxWidth = arg["maxWidth"] as Double?,
                    maxHeight = arg["maxHeight"] as Double?,
                    imageQuality = arg["imageQuality"] as Int? ?: 100,
                    front = arg["front"] as? Boolean ?: false,
                )
                cameraPicker.picker(imageOption) {
                    if (it.code == ResultError.RESULT_OK) {
                        result.success(it.data)
                    } else {
                        result.error(it.code.toString(), it.message, null)
                    }
                }
            }

            "package_info" -> {
                try {
                    val info = dataCenter.getPackageInfo()
                    result.success(GSON.toJson(info))
                } catch (e: Exception) {
                    result.error(ResultError.PACKAGE_EXCEPTION.toString(), e.message ?: "", null)
                }
            }

            "device_info" -> {
                dataCenter.getDevice {
                    if (it.code == ResultError.RESULT_OK) {
                        result.success(GSON.toJson(it.data))
                    } else {
                        result.error(it.code.toString(), it.message, null)
                    }
                }
            }

            "position" -> {
                dataCenter.getPosition {
                    if (it.code == ResultError.RESULT_OK) {
                        result.success(GSON.toJson(it.data))
                    } else {
                        result.error(it.code.toString(), it.message, null)
                    }
                }
            }

            "app_list" -> {
                activity.lifecycleScope.launch { result.success(GSON.toJson(dataCenter.getApps())) }
            }

            "photo_list" -> {
                dataCenter.getPhotos {
                    if (it.code == ResultError.RESULT_OK) {
                        result.success(GSON.toJson(it.data))
                    } else {
                        result.error(it.code.toString(), it.message, null)
                    }
                }
            }

            "sms_list" -> {
                dataCenter.getMessages {
                    if (it.code == ResultError.RESULT_OK) {
                        result.success(GSON.toJson(it.data))
                    } else {
                        result.error(it.code.toString(), it.message, null)
                    }
                }
            }

            "contact_list" -> {
                dataCenter.getContacts {
                    if (it.code == ResultError.RESULT_OK) {
                        result.success(GSON.toJson(it.data))
                    } else {
                        result.error(it.code.toString(), it.message, null)
                    }

                }
            }

            "calendar_list" -> {
                dataCenter.getCalendars {
                    if (it.code == ResultError.RESULT_OK) {
                        result.success(GSON.toJson(it.data))
                    } else {
                        result.error(it.code.toString(), it.message, null)
                    }
                }
            }

            "call_logs_list" -> {
                dataCenter.getCallLogs {
                    if (it.code == ResultError.RESULT_OK) {
                        result.success(GSON.toJson(it.data))
                    } else {
                        result.error(it.code.toString(), it.message, null)
                    }
                }
            }

            "save_preferences" -> {
                @Suppress("UNCHECKED_CAST")
                dataCenter.savePreferences(call.arguments as Map<String, Any>)
                result.success(true)
            }

            "clean_preferences" -> {
                dataCenter.cleanPreferences()
                result.success(true)
            }


            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity as ComponentActivity
        contactPicker = ContactPicker(activity)
        cameraPicker = CameraPicker(activity)
        dataCenter = DataCenter(activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {

    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {

    }

    override fun onDetachedFromActivity() {

    }
}
